src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+    location = []
+    while paths:
+        path_segment = paths.pop(0)
+        location.append(path_segment)
+        # Third time through, service_or_method might be `None`, but we want
+        # to continue walking the path. Everything after the first missing
+        # service/method will be `(location, None)`.
+        if service_or_method:
+
+            # This if condition is handling the case of the first loop here.
+            # we cannot use attribute access on the dal directly because of how
+            # attribute access is deferred with DalMethodRequest objects.
+            if isinstance(service_or_method, dict):
+
+                # If the code being resolved has typo'd a service name, this
+                # could be returning something that is not a service.
+                service_or_method = service_or_method.get(path_segment.name)
+            else:
+                # second time through (and beyond), service or method is a real
+                # service or method, and we do not need to use a special case for
+                # finding the first service.
+                try:
+                    service_or_method = getattr(service_or_method, path_segment.name)
+                    if not service_or_method and callable(service_or_method):
+                        raise DalMethodError(request, path=tuple(location))
+
+                except KeyError:
+                    raise DalMethodError(request, path=tuple(location))
+        else:
+            raise DalMethodError(request, path=tuple(location))
+    return service_or_method
+
+
+class DalMethodResolverMiddleware(MethodMiddleware):
+    """
+    A Method Middleware that resolves deferred dal attribute access to
+    underlying services/methods that has to be called
+    """
+
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        """
+        Args:
+             request: Input from caller
+             handler: Downstream middleware or actual DAL method handler
+        """
+        request.dal_method = resolve(request)
+        return handler(request)
+
+
+def default_handle_dal_method(request: DalMethodRequest):
+    """
+    The default method middleware handler for a DalMethodRequester
+
+    Args:
+        request (DalMethodRequest): The method request context.
+
+    Returns: Mixed
+    """
+    assert request.dal_method, "DAL method not resolved"
+    return request.dal_method(*request.args, **request.kwargs)
+
+
+class InvalidMiddleware(Exception):
+    pass
 
 
 class DataAccessLayer(object):
     """
     Gives you access to a DataManager's services.
     """
 
-    def __init__(self, data_manager):
+    # default middleware classes need to be instantiated before being
+    # passed to the init function because the resulting object is called
+    # directly (__call__).
+    def __init__(
+        self,
+        data_manager,
+        middleware=None,
+        default_middleware=(DalMethodResolverMiddleware(),),
```

Comment:
All instances of `DataAccessLayer` will get the same instance of `DalMethodResolverMiddleware`. Maybe take advantage of the "is class" trick below and pass the type `DalMethodResolverMiddleware`

Reply:
To your point about the same instance thing, we can change this specifically to expect classes, which we instantiate during init so preserve the helpfulness.

Reply:
Sometimes you need to pass in an instance, or at least a factory method. For example, you might need to pass init args to the middleware to configure it.